sled-storage/src/manager.rs

```diff
@@ -590,9 +663,145 @@ impl StorageManager {
+                        requested: config.generation,
+                        current: ledger_data.generation,
+                    });
+                } else if config.generation == ledger_data.generation {
+                    info!(
+                        log,
+                        "Requested generation number matches prior request",
+                    );
+
+                    if ledger_data != &config {
+                        error!(
+                            log,
+                            "Requested configuration changed (with the same generation)";
+                            "generation" => ?config.generation
+                        );
+                        return Err(Error::DatasetConfigurationChanged {
+                            generation: config.generation,
+                        });
+                    }
+                } else {
+                    info!(log, "Request looks newer than prior requests");
+                }
+                ledger
+            }
+            None => {
+                info!(log, "No previously-stored 'requested datasets', creating new ledger");
+                Ledger::<DatasetsConfig>::new_with(
+                    &log,
+                    ledger_paths.clone(),
+                    DatasetsConfig::default(),
+                )
+            }
+        };
+
+        let result = self.datasets_ensure_internal(&log, &config).await;
+
+        let ledger_data = ledger.data_mut();
+        if *ledger_data == config {
+            return Ok(result);
+        }
+        *ledger_data = config;
+        ledger.commit().await?;
+
+        Ok(result)
+    }
+
+    // Attempts to ensure that each dataset exist.
+    //
+    // Does not return an error, because the [DatasetsManagementResult] type
+    // includes details about all possible errors that may occur on
+    // a per-dataset granularity.
+    async fn datasets_ensure_internal(
+        &mut self,
+        log: &Logger,
+        config: &DatasetsConfig,
+    ) -> DatasetsManagementResult {
+        let mut status = vec![];
+        for dataset in &config.datasets {
+            status.push(self.dataset_ensure_internal(log, dataset).await);
+        }
+        DatasetsManagementResult { status }
+    }
+
+    async fn dataset_ensure_internal(
+        &mut self,
+        log: &Logger,
+        config: &DatasetConfig,
+    ) -> DatasetManagementStatus {
+        let log = log.new(o!("name" => config.name.full_name()));
+        info!(log, "Ensuring dataset");
+        let mut status = DatasetManagementStatus {
+            dataset_name: config.name.clone(),
+            err: None,
+        };
+
+        if let Err(err) = self.ensure_dataset(config).await {
+            warn!(log, "Failed to ensure dataset"; "dataset" => ?status.dataset_name, "err" => ?err);
+            status.err = Some(err.to_string());
+        };
+
+        status
+    }
+
+    async fn datasets_list(&mut self) -> Result<DatasetsConfig, Error> {
```

Comment:
May be worth a quick docstring indicating that this always reads the _ledger_, not the actual filesystems on the sled.

Reply:
Sure thing. I'll also update the name to `datasets_config_list` to make it more clear this is reading configuration, rather than parsing `zfs list` output.